CIP-DappConnector/CIP-dapp-connector.md

```diff
@@ -0,0 +1,226 @@
+
+### cardano_check_read_access(): Promise\<bool>
+
+Errors: APIError
+
+Returns true if the full API (`cardano` object) was injected and the dApp has access to the user's wallet, and false otherwise.
+
+
+
+
+## Full API
+
+Upon successful connection via `cardano_request_read_access()`, a javascript object named `cardano` is injected with the following methods. All read-only methods (all but the signing functionality) should not require any as the user has already consented to the dApp reading information about the wallet's state when they agreed to `cardano_request_read_access()`. The remaining methods `cardano.sign_tx()`, `cardano.sign_tx_input()`, `cardano.sign_data()` must request the user's consent in an informative way for each and every API call in order to maintain security.
+
+The API chosen here is for the minimum API necessary for dApp <-> Wallet interactions without convenience functions that don't strictly need the wallet's state to work. The API here is for now also only designed for Shelley's Mary hardfork and thus has NFT support. When Alonzo is released with Plutus support this API will have to be extended.
+
+### cardano.get_utxos(amount: cbor\<value> = undefined, paginate: Paginate = undefined): Promise\<TransactionUnspentOutput[] | undefined>
+
+Errors: `APIError`, `PaginateError`
+
+If `amount` is `undefined`, this shall return a list of all UTXOs (unspent transaction outputs) controlled by the wallet. If `amount` is not `undefined`, this request shall be limited to just the UTXOs that are required to reach the combined ADA/multiasset value target specified in `amount`, and if this cannot be attained, `undefined` shall be returned. The results can be further paginated by `paginate` if it is not `undefined`.
+
+### cardano.get_balance(): Promise\<cbor\<value>>
+
+Errors: `APIError`
+
+Returns the total balance available of the wallet. This is the same as summing the results of `cardano.get_utxos()`, but it is both useful to dApps and likely already maintained by the implementing wallet in a more efficient manner so it has been included in the API as well.
+
+### cardano.get_used_addresses(paginate: Paginate = undefined): Promise\<\<cbor<address>[]>
+
+Errors: `APIError`
+
+Returns a list of all used (included in some on-chain transaction) addresses controlled by the wallet. The results can be further paginated by `paginate` if it is not `undefined`.
+
+### cardano.get_unused_addresses(): Promise\<\<cbor<address>[]>
+
+Errors: `APIError`
+
+Returns a list of unused addresses controlled by the wallet.
+
+### cardano.get_change_address(): Promise\<\<cbor<address>>
+
+Errors: `APIError`
+
+Returns an address owned by the wallet that should be used as a change address to return leftover assets during transaction creation back to the connected wallet. This can be used as a generic receive address as well.
+
+### cardano.sign_tx(tx: cbor\<transaction_body>, metadata: cbor\<auxiliary_data> = undefined): Promise\<cbor\<transaction>>
```

Comment:
1. The auxiliary data are already part of the transaction body so the second argument is redundant.

2. On a second note, I'd suggest going for `witness_tx` instead, for pub-key signatures is only one kind of witnessing method. Transactions containing scripts for instance will require full serialized scripts as witness. Alonzo also brings its own new witnesses with datums and redeemers.  

3. We should also consider reducing the output to the set of witnesses for the whole transaction can then be re-constructed by the DApp. Note that, this is not for "saving bytes", but rather a security argument. There's nothing preventing the wallet on the other end to send back a completely different signed transaction. Returning the full serialized signed transaction encourage developers to forward and submit that blob without any further verification. However, by only returning witnesses and letting the caller assemble the final transaction makes sure that they submit what they intended to get signed.

Reply:
Yes I was suggesting to pass to the wallet the whole transaction object, so that the user could also verify what exactly happens in the transaction and as you said to check for all witnesses since scripts can be involved.
Probably a good idea to just return the witness set.

Reply:
>Returning the full serialized signed transaction encourage developers to forward and submit that blob without any further verification.

Good point. I was looking at it more from the perspective of dApps being the potentially-untrustworthy actor, rather than the wallets as well.

So would we have it something like `cardano.sign_tx(tx: cbor<transaction>, partial_sign: bool = false): Promise<cbor<transaction_witness_set>>`? Or something else? Or have we thought of any other ideas for controlling partial signing?

Reply:
one challenge I can think of with this call is how it would work for cases when the exact metadata is not known in advance - I have in mind the [Catalyst registration](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0015/CIP-0015.md) use case when the wallet is producing the signed metadata at the very moment of signing the transaction (supplying the staking key signature into the metadata object). The `sign_data` call could work for mnemonic-based wallets but hardware wallets don't expose any similar call

We indeed faced a problem of a similar nature in cardano-hw-cli and the recommended flow ended up being letting the user sign a dummy transaction first just to obtain the signed metadata.

Reply:
note that the approach with the dummy transaction to obtain the signed Catalyst registration metadata would not  be viable if the wallet returned via the connector just the transaction witnesses (as the dApp would not have the signature/staking public key to include in the metadata)